In [ ]:
import tensorflow as tf
import math
import numpy as np
import pickle
import gc

In [ ]:
def create_NN():
    #Group I inputs-- 10
    input_size_I = (10,)
    input_size_II = (7,)
    input_size_III = (2,)

    input_layers = []
    output_layers = []

    input_layer_1 = tf.keras.layers.Input(shape = input_size_I, name='inputI')
    input_layers.append(input_layer_1)
    dense_layer_1_1 = tf.keras.layers.Dense(4, activation='relu', name='inputI_dense1')(input_layer_1)
    dense_layer_1_2 = tf.keras.layers.Dense(4, activation='relu', name='inputI_dense2')(dense_layer_1_1)
    output_layers.append(dense_layer_1_2)

    #Group II
    input_layer_2 = tf.keras.layers.Input(shape = input_size_II, name='inputII')
    input_layers.append(input_layer_2)
    dense_layer_2_1 = tf.keras.layers.Dense(4, activation='relu', name='inputII_dense1')(input_layer_2)
    dense_layer_2_2 = tf.keras.layers.Dense(4, activation='relu', name='inputII_dense2')(dense_layer_2_1)
    output_layers.append(dense_layer_2_2)

    #Group III
    for i in range(20):
        input_layer = tf.keras.layers.Input(shape=input_size_III, name=f'input_III_{i}')
        input_layers.append(input_layer)
        dense_layer_1 = tf.keras.layers.Dense(4, activation='relu', name=f'denseIII_1_{i}')(input_layer)
        dense_layer_2 = tf.keras.layers.Dense(8, activation='relu', name=f'denseIII_2_{i}')(dense_layer_1)
        output_layers.append(dense_layer_2)

    concat_output = tf.keras.layers.concatenate(output_layers, name='concatenated_output')

    x1 = tf.keras.layers.Dense(1024, activation='relu', name='dense1cct')(concat_output)
    x2 = tf.keras.layers.Dense(1024, activation='relu', name='dense2cct')(x1)
    x3 = tf.keras.layers.Dense(1024, activation='relu', name='dense3cct')(x2)
    output_layer = tf.keras.layers.Dense(1, activation='sigmoid')(x3)

    model = tf.keras.models.Model(inputs=input_layers, outputs=output_layer)
    return model



en_mod = create_NN()
en_mod.summary()

Model: "model"
__________________________________________________________________________________________________
 Layer (type)                   Output Shape         Param #     Connected to                     
 inputI (InputLayer)            [(None, 10)]         0           []                               
                                                                                                  
 inputII (InputLayer)           [(None, 7)]          0           []                               
                                                                                                  
 input_III_0 (InputLayer)       [(None, 2)]          0           []                               
                                                                                                  
 input_III_1 (InputLayer)       [(None, 2)]          0           []                               
                                                                                              

 denseIII_1_16 (Dense)          (None, 4)            12          ['input_III_16[0][0]']           
                                                                                                  
 denseIII_1_17 (Dense)          (None, 4)            12          ['input_III_17[0][0]']           
                                                                                                  
 denseIII_1_18 (Dense)          (None, 4)            12          ['input_III_18[0][0]']           
                                                                                                  
 denseIII_1_19 (Dense)          (None, 4)            12          ['input_III_19[0][0]']           
                                                                                                  
 inputI_dense2 (Dense)          (None, 4)            20          ['inputI_dense1[0][0]']          
                                                                                                  
 inputII_d

Total params: 2,274,437
Trainable params: 2,274,437
Non-trainable params: 0
__________________________________________________________________________________________________


In [ ]:
en_mod.compile(optimizer='adam',loss="binary_crossentropy",metrics=['accuracy'])

In [ ]:
'''def create_NN2():
    #Group I inputs-- 10
    input_size_I = (10,)
    input_size_II = (7,)
    input_size_III = (2,)

    input_layers = []
    output_layers = []

    input_layer_1 = tf.keras.layers.Input(shape = input_size_I, name='inputI')
    input_layers.append(input_layer_1)
    dense_layer_1_1 = tf.keras.layers.Dense(4, activation='relu', name='inputI_dense1')(input_layer_1)
    dense_layer_1_2 = tf.keras.layers.Dense(4, activation='relu', name='inputI_dense2')(dense_layer_1_1)
    #output_layers.append(dense_layer_1_2)

    #Group II
    input_layer_2 = tf.keras.layers.Input(shape = input_size_II, name='inputII')
    input_layers.append(input_layer_2)
    dense_layer_2_1 = tf.keras.layers.Dense(4, activation='relu', name='inputII_dense1')(input_layer_2)
    dense_layer_2_2 = tf.keras.layers.Dense(4, activation='relu', name='inputII_dense2')(dense_layer_2_1)
    output_layers.append(dense_layer_2_2)

    #Group III
    for i in range(20):
        input_layer = tf.keras.layers.Input(shape=input_size_III, name=f'input_III_{i}')
        input_layers.append(input_layer)
        dense_layer_1 = tf.keras.layers.Dense(4, activation='relu', name=f'denseIII_1_{i}')(input_layer)
        dense_layer_2 = tf.keras.layers.Dense(8, activation='relu', name=f'denseIII_2_{i}')(dense_layer_1)
        output_layers.append(dense_layer_2)

    concat_output = tf.keras.layers.concatenate(output_layers, name='concatenated_output')

    x1 = tf.keras.layers.Dense(1024, activation='relu', name='dense1cct')(concat_output)
    x2 = tf.keras.layers.Dense(1024, activation='relu', name='dense2cct')(x1)
    x3 = tf.keras.layers.Dense(1024, activation='relu', name='dense3cct')(x2)

    concat_output2 = tf.keras.layers.concatenate([dense_layer_1_2,x3], name='concatenated_output2')
    x4 = tf.keras.layers.Dense(1080, activation='relu', name='dense4cct')(concat_output2)
    x5 = tf.keras.layers.Dense(1080, activation='relu', name='dense5cct')(x4)



    output_layer = tf.keras.layers.Dense(1, activation='sigmoid')(x5)

    model = tf.keras.models.Model(inputs=input_layers, outputs=output_layer)
    return model



en_mod2 = create_NN2()
en_mod2.summary()
'''

In [ ]:
#en_mod2.compile(optimizer='adam',loss="binary_crossentropy",metrics=['accuracy'])

In [ ]:
from tensorflow.python.client import device_lib
print(device_lib.list_local_devices())

[name: "/device:CPU:0"
device_type: "CPU"
memory_limit: 268435456
locality {
}
incarnation: 1231847368009064543
xla_global_id: -1
, name: "/device:GPU:0"
device_type: "GPU"
memory_limit: 14236516352
locality {
  bus_id: 1
  links {
  }
}
incarnation: 5183096935519418894
physical_device_desc: "device: 0, name: NVIDIA GeForce RTX 3080 Ti Laptop GPU, pci bus id: 0000:01:00.0, compute capability: 8.6"
xla_global_id: 416903419
]


In [ ]:
output_filepath = "ton_iot_datablockI_1897.pkl"
with open(output_filepath, 'rb') as f:
    groupI_block = pickle.load(f)
groupI_block.shape

(500000, 10)

In [ ]:
groupI_train = groupI_block[0:350000,:]
groupI_valid = groupI_block[350000:425000,:]
groupI_test = groupI_block[425000:,:]
groupI_train.shape,groupI_valid.shape,groupI_test.shape

((350000, 10), (75000, 10), (75000, 10))

In [ ]:
np.min(groupI_train), np.max(groupI_train)

(0.0, 1.0)

In [ ]:
output_filepath = "GroupII_1897_tvals.pkl"

with open(output_filepath, 'rb') as f:
    groupII_block = pickle.load(f)
groupII_block.shape

(500000, 7)

In [ ]:
groupII_train = groupII_block[0:350000,:]
groupII_valid = groupII_block[350000:425000,:]
groupII_test = groupII_block[425000:,:]
groupII_train.shape,groupII_valid.shape,groupII_test.shape

((350000, 7), (75000, 7), (75000, 7))

In [ ]:
np.min(groupII_train), np.max(groupII_train)

(0.0, 1.0)

In [ ]:
output_filepath = "GroupIII_processed_1897"

with open(output_filepath, 'rb') as f:
    groupIII_block = pickle.load(f)
groupIII_block.shape

(500000, 40)

In [ ]:
groupIII_train = groupIII_block[0:350000,:]
groupIII_valid = groupIII_block[350000:425000,:]
groupIII_test = groupIII_block[425000:,:]
groupIII_train.shape,groupIII_valid.shape,groupIII_test.shape

((350000, 40), (75000, 40), (75000, 40))

In [ ]:
np.min(groupIII_train), np.max(groupIII_train)

(nan, nan)

In [ ]:
min_nan_ind = np.where(np.isnan(np.min(groupIII_train, axis=1)))[0]# indices where min value = nan
print(min_nan_ind)

max_nan_ind = np.where(np.isnan(np.max(groupIII_train, axis=1)))[0]# indices where max value = nan
print(max_nan_ind)

nan_ind = np.union1d(min_nan_ind, max_nan_ind) #combined
print("Rows with nan minimum or maximum value:", nan_ind)


[1]
[1]
Rows with nan minimum or maximum value: [1]


In [ ]:
groupIII_train = np.delete(groupIII_train,1, axis = 0)
groupIII_train.shape

(349999, 40)

In [ ]:
groupI_train = np.delete(groupI_train,1,axis=0)
groupII_train = np.delete(groupII_train,1,axis=0)
groupI_train.shape,groupII_train.shape

((349999, 10), (349999, 7))

In [ ]:
del groupI_block
del groupII_block
del groupIII_block
gc.collect()

491

In [ ]:
labels_block = np.load('labels_seeded_1897.npy')
labels_block.shape

(500000,)

In [ ]:
the_inputs = [groupI_train,groupII_train]
the_valids = [groupI_valid,groupII_valid]
the_tests = [groupI_test,groupII_test]

In [ ]:
labels_tr = labels_block[0:350000,]
labels_va = labels_block[350000:425000,]
labels_te = labels_block[425000:,]
labels_tr.shape,labels_va.shape,labels_te.shape

((350000,), (75000,), (75000,))

In [ ]:
labels_tr = np.delete(labels_tr,1, axis = 0)
labels_tr.shape

(349999,)

In [ ]:
def split_20(incoming:np.array):
    split_arrays = np.split(incoming,20,axis=1)
    arrays_dict = {}
    for i, arr in enumerate(split_arrays):
        arrays_dict[f'arr{i:02d}'] = arr
    return arrays_dict

In [ ]:
the_dict_tr = split_20(groupIII_train)
the_dict_va = split_20(groupIII_valid)
the_dict_te = split_20(groupIII_test)

In [ ]:
for i in range(20):
    the_inputs.append(the_dict_tr[f'arr{i:02d}'])
    the_valids.append(the_dict_va[f'arr{i:02d}'])
    the_tests.append(the_dict_te[f'arr{i:02d}'])


In [ ]:
len(the_inputs), len(the_valids), len(the_tests)

(22, 22, 22)

In [ ]:
#for en_mod_s
EPOCHS = 200
batchSize = 64
early_stopping = tf.keras.callbacks.EarlyStopping(monitor='val_accuracy', patience=5, restore_best_weights=True)
history = en_mod.fit(
    x=the_inputs,
    y=labels_tr,
    batch_size=batchSize,
    epochs=EPOCHS,
    validation_data=(the_valids, labels_va),
    callbacks=[early_stopping],
    verbose=1
)

Epoch 1/200
5469/5469 [==============================] - 133s 23ms/step - loss: 0.1248 - accuracy: 0.9568 - val_loss: 0.1196 - val_accuracy: 0.9597
Epoch 2/200
5469/5469 [==============================] - 135s 25ms/step - loss: 0.0844 - accuracy: 0.9693 - val_loss: 0.0752 - val_accuracy: 0.9709
Epoch 3/200
5469/5469 [==============================] - 127s 23ms/step - loss: 0.0723 - accuracy: 0.9729 - val_loss: 0.0685 - val_accuracy: 0.9754
Epoch 4/200
5469/5469 [==============================] - 128s 23ms/step - loss: 0.0682 - accuracy: 0.9744 - val_loss: 0.0808 - val_accuracy: 0.9740
Epoch 5/200
5469/5469 [==============================] - 130s 24ms/step - loss: 0.0637 - accuracy: 0.9762 - val_loss: 0.0577 - val_accuracy: 0.9785
Epoch 6/200
5469/5469 [==============================] - 130s 24ms/step - loss: 0.0638 - accuracy: 0.9767 - val_loss: 0.0566 - val_accuracy: 0.9786
Epoch 7/200
5469/5469 [==============================] - 131s 24ms/step - loss: 0.0592 - accuracy: 0.9779 - val_

In [ ]:
output_filepath = "model_III_10_7_20_uniform_concat.pkl"
en_mod.save(output_filepath)


INFO:tensorflow:Assets written to: model_III_10_7_20_uniform_concat.pkl\assets


INFO:tensorflow:Assets written to: model_III_10_7_20_uniform_concat.pkl\assets


In [ ]:
predictions = en_mod.predict(the_tests, verbose=1)
predicted_labels = (predictions > 0.5).astype(int)
from sklearn.metrics import classification_report
print(classification_report(labels_te, predicted_labels))

1963/2344 [========================>.....] - ETA: 4s

In [ ]:
EPOCHS = 200
batchSize = 64
early_stopping = tf.keras.callbacks.EarlyStopping(monitor='val_accuracy', patience=5, restore_best_weights=True)
history = en_mod.fit(
    x=the_inputs,
    y=labels_tr,
    batch_size=batchSize,
    epochs=EPOCHS,
    validation_data=(the_valids, labels_va),
    callbacks=[early_stopping],
    verbose=1
)

2344/2344 [==============================] - 25s 10ms/step
              precision    recall  f1-score   support

           0       0.99      0.97      0.98     27071
           1       0.98      0.99      0.99     47929

    accuracy                           0.98     75000
   macro avg       0.99      0.98      0.98     75000
weighted avg       0.98      0.98      0.98     75000
